pkg/controller/controller.go

```diff
@@ -0,0 +1,131 @@
+	HandleNext func() (interface{}, error)
+
+	// RetryManager is fed the handled resource if Handle returns a Retryable
+	// error. If Handle returns no error, the RetryManager is asked to forget
+	// the resource.
+	RetryManager RetryManager
+
+	// Stop is used to shut down the controller.
+	Stop <-chan struct{}
+}
+
+// Run will call Handle until a message is received on Stop. If Handle returns
+// a retryable error, the handled resource is passed to the RetryManager. If
+// no error is returned from Handle, the RetryManager is asked to forget the
+// processed resource.
+func (c *RetryController) Run() {
+	go kutil.Until(func() {
+		resource, err := c.HandleNext()
+		if err != nil {
+			if _, isRetryable := err.(*RetryableError); isRetryable {
+				c.RetryManager.Retry(resource)
+				return
+			}
+		}
+		c.RetryManager.Forget(resource)
+	}, 0, c.Stop)
+}
+
+// RetryableError is an error that is recoverable and which should
+// signal another attempt at processing a resource.
+type RetryableError struct {
+	err error
+}
+
+func (e *RetryableError) Error() string {
+	return e.err.Error()
+}
+
+func NewRetryableError(err error) *RetryableError {
+	return &RetryableError{err}
+}
+
+// RetryManager knows how to retry processing of a resource, and how to forget
+// a resource it may be tracking the state of.
+type RetryManager interface {
+	// Retry will cause resource processing to be retried (for example, by
+	// requeueing resource)
+	Retry(resource interface{})
+
+	// Forget will cause the manager to erase all prior knowledge of resource
+	// and reclaim internal resources associated with state tracking of
+	// resource.
+	Forget(resource interface{})
+}
+
+// QueueRetryManager retries a resource by re-queueing it into a FIFO up to
+// MaxRetries number of times.
+type QueueRetryManager struct {
+	// queue is where resources are re-queued.
+	queue FIFO
+
+	// keyFunc is used to index resources.
+	keyFunc kcache.KeyFunc
+
+	// maxRetries is the total number of attempts to requeue an individual
+	// resource before giving up.
+	maxRetries int
+
+	// retries maps resources to their total retry count.
+	retries map[string]int
+}
+
+// FIFO is a narrow abstraction of a cache.FIFO.
+type FIFO interface {
+	AddIfNotPresent(interface{}) error
+}
+
+// NewQueueRetryManager safely creates a new QueueRetryManager.
+func NewQueueRetryManager(queue FIFO, keyFunc kcache.KeyFunc, maxRetries int) *QueueRetryManager {
+	return &QueueRetryManager{
+		queue:      queue,
+		keyFunc:    keyFunc,
+		maxRetries: maxRetries,
+		retries:    make(map[string]int),
+	}
+}
+
+// Retry will enqueue resource until MaxRetries for that resource has been
+// exceeded, at which point resource will be forgotten and no longer retried.
+func (r *QueueRetryManager) Retry(resource interface{}) {
+	id, _ := r.keyFunc(resource)
+
+	if _, exists := r.retries[id]; !exists {
+		r.retries[id] = 0
+	}
+	tries := r.retries[id]
+
+	if tries < r.maxRetries {
```

Comment:
Done. Also, I hate using a constructor func rather than field initialization for this thing, but then the map would have to be lazily initialized. Ugh!